### PR TITLE
fix(xsnap): Account for TypedArray and subarrays in Text shim

### PR DESCRIPTION
**Pull request by kriskowal**
refs: This change is necessary for integration of Zip archives #3346

At this time, XS does not provide a node-nor-web-compatible `TextEncoder` or `TextDecoder` so `xsnap` provides a shim. The shim is written in terms of XS's proprietary `String.fromArrayBuffer` and `ArrayBuffer.fromString` which do not yet take into account TypedArrays (only accepting ArrayBuffer, strictly) and do not account for subarray views that are not aligned at byteOffset zero. This change adds a test, and a temporary fix. When we next update Moddable/XS, the fix can be removed.

Caveat: the underlying XS API still truncates strings at the first encountered C null terminator which is necessary for continued compatibility with other tools in the Moddable ecosystem. Moddable may mitigate this by adding a module for a web-and-node-compatible `TextEncoder` and `TextDecoder` in a future version, but the limitation is currently bearable.

### CODE REVIEW

File: packages/xsnap/lib/text-shim.js [hunk 1]
```diff
@@ -7,14 +7,21 @@ const { fromString } = ArrayBuffer;
...
+  encode(text) {
+    return new Uint8Array(fromString(text));
   }
 }
 
 class TextDecoder {
-  decode(bs) {
-    return fromArrayBuffer(bs);
+  decode(bytes) {
+    // TODO: the following condition can be removed in a future update of XS.
```
Comment (dckc) [inline packages/xsnap/lib/text-shim.js:17]: I suppose it's OK if this TODO never gets done, so this isn't critical, but...

What are the norms around TODOs around here? Should this have a pointer to an issue? In the moddable repo? Or should it be down-graded from TODO to NOTE before merging or something?

File: packages/xsnap/lib/text-shim.js [hunk 2]
```diff
@@ -7,14 +7,21 @@ const { fromString } = ArrayBuffer;
 const { fromArrayBuffer } = String;
 
 class TextEncoder {
-  encode(s) {
-    return new Uint8Array(fromString(s));
+  encode(text) {
```
Comment (dckc) [inline packages/xsnap/lib/text-shim.js:10]: Changing `s` to `text` seems like a bit of a drive-by. I suppose it's a welcome change, but I prefer to see at least a mention of this sort of thing in the commit message.

File: packages/xsnap/test/test-boot-lockdown.js [hunk 3]
```diff
@@ -80,3 +80,23 @@ test('SES deep stacks work on xsnap', async t => {
...
+  const opts = options();
+  const vat = xsnap(opts);
+  await vat.evaluate(bootScript);
+  await vat.evaluate(`
+    const decoder = new TextDecoder();
+    const encoder = new TextEncoder();
+    const send = msg => issueCommand(encoder.encode(JSON.stringify(msg)).buffer);
+
+    const string = '0123456789';
+    const bytes = encoder.encode(string).subarray(1, 4);
```
Comment (dckc) [inline packages/xsnap/test/test-boot-lockdown.js:82]: I'm a little fuzzy on how this test would fail without the fix in this PR. I suppose I should test it, but I'll take the risk that I may need to dig into this later instead.

**Review by dckc — APPROVED** [review]
good stuff!

File: packages/xsnap/test/test-boot-lockdown.js (hunk 3 above)
Comment (kriskowal) [reply packages/xsnap/test/test-boot-lockdown.js:82]: The test failed previously for a couple reasons. The first was that String.fromArrayBuffer threw an exception if you passed a Uint8Array (which is a TypedArray view of an ArrayBuffer). The Uint8Array has a `buffer` property you can use instead, but that would fail because String.fromArrayBuffer assumes the data to decode starts at byteOffset 0, but a TypeArray view can be a view into a window at a non-zero offset. So, this test ensures that the `subarray(1, 4)` causes truncates the buffer at an offset and that carries over as as equivalent to the change to the resulting string.

The latest XS already addresses both of these problems by making `String.fromArrayBuffer` accept `TypeArray` views into array buffers and also taking into account their `byteOffset` and `byteLength`. It does this without resorting to a memcopy, which is strictly better than this work-around.

File: packages/xsnap/lib/text-shim.js (hunk 1 above)
Comment (kriskowal) [reply packages/xsnap/lib/text-shim.js:17]: Added a link. I don’t know the norms, but I think tracking with an issue should suffice. Thanks for the suggestion. https://github.com/Agoric/agoric-sdk/issues/3362

File: packages/xsnap/lib/text-shim.js (hunk 2 above)
Comment (kriskowal) [reply packages/xsnap/lib/text-shim.js:10]: That’s fair. I’ve split the commit into the body and a follow-up: `refactor(xsnap): Expand terse argument name in text shim for local consistency`
